File: ws/collahuazo/u2/ws13GradesStudentAppCLI/JavaScript/view/gradesStudentApp.js
const readline = require("readline");
const StudentController = require("../controller/studentController");

class GradesStudentApp {
  constructor() {
    this.controller = new StudentController();
    this.rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });
  }

  ask(question) {
    return new Promise((resolve) => {
      this.rl.question(question, resolve);
    });
  }

  async main() {
    while (true) {
      console.log("\n==================================");
      console.log("       STUDENT GRADES SYSTEM      ");
      console.log("==================================");
      console.log("1. Register Student");
      console.log("2. Input Grades (Option A)");
      console.log("3. Grades and Students Report");
      console.log("4. Exit");

      const optionInput = await this.ask("Select an option: ");
      const option = Number(optionInput.trim());

      if (Number.isNaN(option)) {
        console.log("Invalid input. Please enter a number.");
        continue;
      }

      if (option === 1) {
        const studentId = (await this.ask("Enter ID: ")).trim();
        const firstName = (await this.ask("Enter First Name: ")).trim();
        const lastName = (await this.ask("Enter Last Name: ")).trim();
        const phone = (await this.ask("Enter Phone Number: ")).trim();

        this.controller.registerStudent(studentId, firstName, lastName, phone);
        console.log("Student registered successfully!");
      } else if (option === 2) {
        const students = this.controller.getAllStudents();
        console.log("\n--- Registered Students ---");

        if (students.length === 0) {
          console.log("No students found. Register one first.");
          continue;
        }

        for (const student of students) {
          console.log(`ID: ${student.id} | Name: ${student.firstName} ${student.lastName}`);
        }

        const targetId = (await this.ask("\nEnter the ID of the student to add grades: ")).trim();
        if (!this.controller.findStudentById(targetId)) {
          console.log("Student ID not found.");
          continue;
        }

        const temporaryGrades = [];
        let inputError = false;

        for (let i = 1; i <= 3; i += 1) {
          const gradeInput = (await this.ask(`Enter grade ${i} (0 - 20): `)).trim();
          const gradeValue = Number(gradeInput);
          if (Number.isNaN(gradeValue)) {
            console.log("Invalid grade format. Input cancelled.");
            inputError = true;
            break;
          }
          temporaryGrades.push(gradeValue);
        }

        if (!inputError) {
          if (this.controller.addGradesToStudent(targetId, temporaryGrades)) {
            console.log("All 3 grades added successfully!");
          } else {
            console.log("Grades could not be saved due to validation rules.");
          }
        }
      } else if (option === 3) {
        const students = this.controller.getAllStudents();
        console.log("\n+------------+-----------------+-----------------+---------------------------+---------+-----------+");
        console.log("| ID         | First Name     | Last Name      | Grades                    | Average | Status    |");
        console.log("+------------+-----------------+-----------------+---------------------------+---------+-----------+");

        if (students.length === 0) {
          console.log("| No data available                                                                 |");
        } else {
          for (const student of students) {
            const studentGrades = this.controller.getGradesByStudent(student.id);
            const avg = this.controller.calculateAverage(student.id);
            const status = avg >= 14.0 ? "Pass" : "No pass";
            const gradesStr = JSON.stringify(studentGrades);
            console.log(`| ${student.id.padEnd(10)} | ${student.firstName.padEnd(15)} | ${student.lastName.padEnd(15)} | ${gradesStr.padEnd(25)} | ${avg.toFixed(2).padEnd(7)} | ${status.padEnd(9)} |`);
          }
        }

        console.log("+------------+-----------------+-----------------+---------------------------+---------+-----------+");
      } else if (option === 4) {
        console.log("Exiting the program... Goodbye!");
        break;
      } else {
        console.log("Invalid option. Try again.");
      }
    }

    this.rl.close();
  }
}

if (require.main === module) {
  const app = new GradesStudentApp();
  app.main();
}

module.exports = GradesStudentApp;

File: ws/collahuazo/u2/ws13GradesStudentAppCLI/Python/view/grades_student_app.py
from controller.student_controller import StudentController


class GradesStudentApp:
    def __init__(self):
        self.controller = StudentController()

    def main(self):
        while True:
            print("\n==================================")
            print("       STUDENT GRADES SYSTEM      ")
            print("==================================")
            print("1. Register Student")
            print("2. Input Grades (Option A)")
            print("3. Grades and Students Report")
            print("4. Exit")

            option = input("Select an option: ").strip()
            if not option.isdigit():
                print("Invalid input. Please enter a number.")
                continue

            option = int(option)

            if option == 1:
                student_id = input("Enter ID: ").strip()
                first_name = input("Enter First Name: ").strip()
                last_name = input("Enter Last Name: ").strip()
                phone = input("Enter Phone Number: ").strip()
                self.controller.register_student(student_id, first_name, last_name, phone)
                print("Student registered successfully!")

            elif option == 2:
                students = self.controller.get_all_students()
                print("\n--- Registered Students ---")
                if not students:
                    print("No students found. Register one first.")
                    continue

                for student in students:
                    print(f"ID: {student.id} | Name: {student.first_name} {student.last_name}")

                target_id = input("\nEnter the ID of the student to add grades: ").strip()
                if self.controller.find_student_by_id(target_id) is None:
                    print("Student ID not found.")
                    continue

                temporary_grades = []
                input_error = False
                for i in range(1, 4):
                    grade_input = input(f"Enter grade {i} (0 - 20): ").strip()
                    try:
                        grade_value = float(grade_input)
                        temporary_grades.append(grade_value)
                    except ValueError:
                        print("Invalid grade format. Input cancelled.")
                        input_error = True
                        break

                if not input_error:
                    if self.controller.add_grades_to_student(target_id, temporary_grades):
                        print("All 3 grades added successfully!")
                    else:
                        print("Grades could not be saved due to validation rules.")

            elif option == 3:
                students = self.controller.get_all_students()
                print("\n+------------+-----------------+-----------------+---------------------------+---------+-----------+")
                print("| %-10s | %-15s | %-15s | %-25s | %-7s | %-9s |" % ("ID", "First Name", "Last Name", "Grades", "Average", "Status"))
                print("+------------+-----------------+-----------------+---------------------------+---------+-----------+")

                if not students:
                    print("| %-92s |" % "No data available")
                else:
                    for student in students:
                        student_grades = self.controller.get_grades_by_student(student.id)
                        avg = self.controller.calculate_average(student.id)
                        status = "Pass" if avg >= 14.0 else "No pass"
                        grades_str = str(student_grades)
                        print("| %-10s | %-15s | %-15s | %-25s | %-7.2f | %-9s |" % (
                            student.id,
                            student.first_name,
                            student.last_name,
                            grades_str,
                            avg,
                            status,
                        ))

                print("+------------+-----------------+-----------------+---------------------------+---------+-----------+")

            elif option == 4:
                print("Exiting the program... Goodbye!")
                break

            else:
                print("Invalid option. Try again.")


if __name__ == "__main__":
    GradesStudentApp().main()
